File: integrasjon/src/test/java/no/nav/melosys/integrasjon/medl/behandle/MedlPeriodeKonverterTest.java
package no.nav.melosys.integrasjon.medl.behandle;

import no.nav.melosys.domain.Lovvalgsperiode;
import no.nav.melosys.domain.kodeverk.Folketrygdloven_kap2_bestemmelser;
import no.nav.melosys.domain.kodeverk.LovvalgBestemmelse;
import no.nav.melosys.domain.kodeverk.Trygdedekninger;
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.Lovvalgbestemmelser_883_2004;
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.Lovvalgbestemmelser_987_2009;
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.Tilleggsbestemmelser_883_2004;
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.trygdeavtale.Lovvalgsbestemmelser_trygdeavtale_gb;
import no.nav.melosys.exception.TekniskException;
import no.nav.melosys.integrasjon.medl.DekningMedl;
import no.nav.melosys.integrasjon.medl.GrunnlagMedl;
import no.nav.melosys.integrasjon.medl.MedlPeriodeKonverter;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;

class MedlPeriodeKonverterTest {

    @Test
    void tilGrunnlagMedltype() {
        assertThat(MedlPeriodeKonverter.tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_2))
            .isEqualTo(GrunnlagMedl.FO_12_2);
        assertThat(MedlPeriodeKonverter.tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART16_1))
            .isEqualTo(GrunnlagMedl.FO_16);
        assertThat(MedlPeriodeKonverter.tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART16_2))
            .isEqualTo(GrunnlagMedl.FO_16);
        assertThat(MedlPeriodeKonverter.tilGrunnlagMedltype(Lovvalgsbestemmelser_trygdeavtale_gb.UK_ART7_3))
            .isEqualTo(GrunnlagMedl.Storbrit_NIrland_7_3);
        assertThat(MedlPeriodeKonverter.tilGrunnlagMedltype(Lovvalgbestemmelser_987_2009.FO_987_2009_ART14_11))
            .isEqualTo(GrunnlagMedl.FO_987_2009_14_11);

        assertThatExceptionOfType(TekniskException.class)
            .isThrownBy(() -> MedlPeriodeKonverter.tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART11_1))
            .withMessageContaining("støttes ikke i MEDL");
    }

    @Test
    void tilGrunnlagMedltype_Ftrl() {
        assertThat(MedlPeriodeKonverter.tilGrunnlagMedltype(Folketrygdloven_kap2_bestemmelser.FTRL_KAP2_2_8)).isEqualTo(GrunnlagMedl.FTL_2_8);

        assertThatExceptionOfType(TekniskException.class)
            .isThrownBy(() -> MedlPeriodeKonverter.tilGrunnlagMedltype(Folketrygdloven_kap2_bestemmelser.FTRL_KAP2_2_1))
            .withMessageContaining("Folketrygdloven bestemmelse støttes ikke. Kode: FTRL_KAP2_2_1");
    }

    @Test
    void tilLovvalgBestemmelse() {
        assertThat(MedlPeriodeKonverter.tilLovvalgBestemmelse(GrunnlagMedl.FO_12_2))
            .isEqualTo(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_2);
        assertThat(MedlPeriodeKonverter.tilLovvalgBestemmelse(GrunnlagMedl.FO_16))
            .isEqualTo(Lovvalgbestemmelser_883_2004.FO_883_2004_ART16_1);

        assertThatExceptionOfType(TekniskException.class)
            .isThrownBy(() -> MedlPeriodeKonverter.tilLovvalgBestemmelse(GrunnlagMedl.MEDFT))
            .withMessageContaining("GrunnlagMedlKode er ukjent");
    }

    @Test
    void hentFellesKodeForDekningtype() {
        Trygdedekninger trygdeDekning = Trygdedekninger.UTEN_DEKNING;
        DekningMedl dekningMedl = MedlPeriodeKonverter.tilMedlTrygdeDekningEos(trygdeDekning);
        assertThat(dekningMedl).isEqualTo(DekningMedl.UNNTATT);
    }

    @Test
    void hentLovvalgbestemmelse() {
        assertThat(MedlPeriodeKonverter.hentLovvalgBestemmelse(lovvalgsperiode(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1, null)))
            .isEqualTo(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1);
        assertThat(MedlPeriodeKonverter.hentLovvalgBestemmelse(lovvalgsperiode(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1, Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_5)))
            .isEqualTo(Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_5);
        assertThat(MedlPeriodeKonverter.hentLovvalgBestemmelse(lovvalgsperiode(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1, Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_4_1)))
            .isEqualTo(Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_4_1);
    }

    private Lovvalgsperiode lovvalgsperiode(LovvalgBestemmelse lovvalgBestemmelse, LovvalgBestemmelse tilleggBestemmelse) {
        var lovvalgsperiode = new Lovvalgsperiode();
        lovvalgsperiode.setBestemmelse(lovvalgBestemmelse);
        lovvalgsperiode.setTilleggsbestemmelse(tilleggBestemmelse);
        return lovvalgsperiode;
    }
}

File: integrasjon/src/test/kotlin/no/nav/melosys/integrasjon/medl/behandle/MedlPeriodeKonverterTest.kt
package no.nav.melosys.integrasjon.medl.behandle

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import no.nav.melosys.domain.Lovvalgsperiode
import no.nav.melosys.domain.kodeverk.Folketrygdloven_kap2_bestemmelser
import no.nav.melosys.domain.kodeverk.LovvalgBestemmelse
import no.nav.melosys.domain.kodeverk.Trygdedekninger
import no.nav.melosys.domain.kodeverk.Vertslandsavtale_bestemmelser
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.*
import no.nav.melosys.domain.kodeverk.lovvalgsbestemmelser.trygdeavtale.Lovvalgsbestemmelser_trygdeavtale_gb
import no.nav.melosys.exception.TekniskException
import no.nav.melosys.integrasjon.medl.DekningMedl
import no.nav.melosys.integrasjon.medl.GrunnlagMedl
import no.nav.melosys.integrasjon.medl.MedlPeriodeKonverter.hentLovvalgBestemmelse
import no.nav.melosys.integrasjon.medl.MedlPeriodeKonverter.tilGrunnlagMedltype
import no.nav.melosys.integrasjon.medl.MedlPeriodeKonverter.tilLovvalgBestemmelse
import no.nav.melosys.integrasjon.medl.MedlPeriodeKonverter.tilMedlTrygdeDekning
import no.nav.melosys.integrasjon.medl.MedlPeriodeKonverter.tilMedlTrygdedekningForFtrl
import org.junit.jupiter.api.Test

internal class MedlPeriodeKonverterTest {
    @Test
    fun tilGrunnlagMedltype() {
        tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_2) shouldBe GrunnlagMedl.FO_12_2
        tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART16_1) shouldBe GrunnlagMedl.FO_16
        tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART16_2) shouldBe GrunnlagMedl.FO_16
        tilGrunnlagMedltype(Lovvalgsbestemmelser_trygdeavtale_gb.UK_ART7_3) shouldBe GrunnlagMedl.STORBRIT_NIRLAND_7_3
        tilGrunnlagMedltype(Lovvalgbestemmelser_987_2009.FO_987_2009_ART14_11) shouldBe GrunnlagMedl.FO_987_2009_14_11

        shouldThrow<TekniskException> {
            tilGrunnlagMedltype(Lovvalgbestemmelser_883_2004.FO_883_2004_ART11_1)
        }.message shouldContain "støttes ikke i MEDL"
    }

    @Test
    fun tilGrunnlagMedltype_FtrlOgSpesielleGrupper() {
        tilGrunnlagMedltype(Folketrygdloven_kap2_bestemmelser.FTRL_KAP2_2_8_FØRSTE_LEDD_A) shouldBe GrunnlagMedl.FTL_2_8_1_LEDD_A
        tilGrunnlagMedltype(Vertslandsavtale_bestemmelser.TILLEGGSAVTALE_NATO) shouldBe GrunnlagMedl.TILLEGGSAVTALE_NATO
        tilGrunnlagMedltype(Folketrygdloven_kap2_bestemmelser.FTRL_KAP2_2_8_FØRSTE_LEDD_B) shouldBe GrunnlagMedl.FTL_2_8_1_LEDD_B
        tilGrunnlagMedltype(Vertslandsavtale_bestemmelser.DET_INTERNASJONALE_BARENTSSEKRETARIATET_ART14) shouldBe GrunnlagMedl.DET_INTERNASJONALE_BARENTSSEKRETARIATET_14

        shouldThrow<TekniskException> {
            tilGrunnlagMedltype(Folketrygdloven_kap2_bestemmelser.FTRL_KAP2_2_5)
        }.message shouldContain "Folketrygdloven bestemmelse støttes ikke."
    }

    @Test
    fun tilLovvalgBestemmelse() {
        tilLovvalgBestemmelse(GrunnlagMedl.FO_12_2) shouldBe Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_2
        tilLovvalgBestemmelse(GrunnlagMedl.FO_16) shouldBe Lovvalgbestemmelser_883_2004.FO_883_2004_ART16_1

        shouldThrow<TekniskException> {
            tilLovvalgBestemmelse(GrunnlagMedl.MEDFT)
        }.message shouldContain "GrunnlagMedlKode er ukjent"
    }

    @Test
    fun tilMedlTrygdedekningForFtrl() {
        tilMedlTrygdedekningForFtrl(Trygdedekninger.TILLEGGSAVTALE_NATO_HELSEDEL) shouldBe DekningMedl.TILLEGSAVTALE_NATO_DEKNING

        shouldThrow<TekniskException> {
            tilMedlTrygdedekningForFtrl(Trygdedekninger.UNNTATT_USA_5_2_G)
        }.message shouldContain "Dekningstype støttes ikke for FTRL: ${Trygdedekninger.UNNTATT_USA_5_2_G.kode}"
    }

    @Test
    fun hentFellesKodeForDekningtype() {
        val trygdeDekning = Trygdedekninger.UTEN_DEKNING
        val dekningMedl = tilMedlTrygdeDekning(trygdeDekning)
        dekningMedl shouldBe DekningMedl.UNNTATT
    }

    @Test
    fun hentLovvalgbestemmelse() {
        hentLovvalgBestemmelse(lagLovvalgsperiode(Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1, null)) shouldBe Lovvalgbestemmelser_883_2004
            .FO_883_2004_ART12_1

        hentLovvalgBestemmelse(
            lagLovvalgsperiode(
                Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1,
                Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_5
            )
        ) shouldBe Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1

        hentLovvalgBestemmelse(
            lagLovvalgsperiode(
                Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1,
                Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_4_1
            )
        ) shouldBe Lovvalgbestemmelser_883_2004.FO_883_2004_ART12_1

        hentLovvalgBestemmelse(
            lagLovvalgsperiode(
                Lovvalgbestemmelser_konv_efta_storbritannia.KONV_EFTA_STORBRITANNIA_ART13_3A,
                Tilleggsbestemmelser_konv_efta_storbritannia.KONV_EFTA_STORBRITANNIA_ART13_4_1
            )
        ) shouldBe Tilleggsbestemmelser_konv_efta_storbritannia.KONV_EFTA_STORBRITANNIA_ART13_4_1

        hentLovvalgBestemmelse(
            lagLovvalgsperiode(
                Lovvalgbestemmelser_883_2004.FO_883_2004_ART11_3A,
                Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_4_1
            )
        ) shouldBe Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_4_1

        hentLovvalgBestemmelse(
            lagLovvalgsperiode(
                Lovvalgbestemmelser_883_2004.FO_883_2004_ART11_3A,
                Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_5
            )
        ) shouldBe Tilleggsbestemmelser_883_2004.FO_883_2004_ART11_5
    }

    private fun lagLovvalgsperiode(lovvalgBestemmelse: LovvalgBestemmelse, tilleggBestemmelse: LovvalgBestemmelse?) =
        Lovvalgsperiode().apply {
            bestemmelse = lovvalgBestemmelse
            tilleggsbestemmelse = tilleggBestemmelse
        }
}
